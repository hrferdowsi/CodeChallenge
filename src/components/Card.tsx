import React from 'react';
import {Box, ButtonBase, Grid, styled, Rating, Typography} from "@mui/material";
import CircleOutlinedIcon from '@mui/icons-material/CircleOutlined';
import CircleSharpIcon from '@mui/icons-material/CircleSharp';


const Img = styled('img')({
    margin: 'auto',
    display: 'block',
    maxWidth: '100%',
    maxHeight: '100%',
});

interface cardsProps {
    data: Accommodation;
}

const Card = ({data}: cardsProps): JSX.Element => {

    const previewURL = data.property.previewImage.url ?? "//:0";

    const AccRates: JSX.Element = (
        data.property.rating.ratingType === "self" ?
            <Rating
                name="read-only"
                value={data.property.rating.ratingValue}
                precision={0.5}
                icon={<CircleSharpIcon fontSize="inherit"/>}
                emptyIcon={<CircleOutlinedIcon fontSize="inherit"/>}
                readOnly
                size={"small"}
            /> :
            <Rating
                name="read-only"
                value={data.property.rating.ratingValue}
                precision={0.5}
                readOnly
                size={"small"}
            />
    )

    return (
        <Grid container spacing={1}>
            <Grid item>
                <ButtonBase sx={{width: 145, height: 125}}>
                    <Img alt={data.property.previewImage.caption} src={previewURL} loading={"lazy"}/>
                    <Box
                        sx={{
                            fontSize: 12,
                            bgcolor: 'white',
                            color: 'red',
                            p: 0.5,
                            position: 'absolute',
                            top: 20,
                            left: 0,
                            zIndex: 'tooltip',
                        }}
                    >
                        {data.offer.promotion.title}
                    </Box>
                </ButtonBase>
            </Grid>
            <Grid item xs={11} sm container>
                <Grid item xs container direction="column" spacing={2}>
                    <Grid item xs>
                        <Typography gutterBottom variant="body1" component="div">
                            <b>{data.property.title}</b> {AccRates}
                        </Typography>
                        <Typography variant="subtitle1" gutterBottom>
                            {data.property.address.join("")}
                        </Typography>
                        <Typography variant="body2" sx={{color: "red"}} gutterBottom>
                            <u>{data.offer.name}</u>
                        </Typography>
                    </Grid>
                    <Grid item>
                        <Typography variant="body2" sx={{color: "green"}} gutterBottom>
                            {data.offer.cancellationOption.cancellationType === "FREE_CANCELLATION" && "Free cancellation"}
                        </Typography>
                    </Grid>
                </Grid>
                <Grid item>
                    <Typography variant="subtitle1" component="div">
                        1 night total (AUD)
                    </Typography>
                    <Typography variant="subtitle1" component="div">
                        ${data.offer.displayPrice.amount}
                    </Typography>
                    {data.offer.savings && (
                        <Typography variant="subtitle1" component="div">
                            Save ${data.offer.savings.amount}~
                        </Typography>
                    )}
                </Grid>
            </Grid>
        </Grid>
    );
};

export default Card;