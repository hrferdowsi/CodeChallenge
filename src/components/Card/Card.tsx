import React from 'react';
import {Box, Link, Stack, ButtonBase, Grid, Rating, Typography} from "@mui/material";
import CircleOutlinedIcon from '@mui/icons-material/CircleOutlined';
import CircleSharpIcon from '@mui/icons-material/CircleSharp';
import {Img, cardImageSize, ImgRibbonStyle, SPACING, SPACING_TWO} from '../../styles';


interface CardsProps {
    data: Accommodation;
}

const Card: React.FC<CardsProps> = ({data}) => {

    const previewURL = data.property.previewImage.url ?? "//:0";

    const AccRates: JSX.Element = (
        data.property.rating.ratingType === "self" ?
            <Rating
                name="read-only"
                value={data.property.rating.ratingValue}
                precision={0.5}
                icon={<CircleSharpIcon fontSize="inherit"/>}
                emptyIcon={<CircleOutlinedIcon fontSize="inherit"/>}
                readOnly
                size={"small"}
                data-testid="rating"
            /> :
            <Rating
                name="read-only"
                value={data.property.rating.ratingValue}
                precision={0.5}
                readOnly
                size={"small"}
                data-testid="rating"
            />
    )

    return (
        <Grid container spacing={SPACING}>
            <Grid item>
                <ButtonBase href="#" sx={cardImageSize} data-testid="hotel_picture">
                    <Img alt={data.property.previewImage.caption} src={previewURL} loading={"lazy"}/>
                    <Box sx={ImgRibbonStyle}>
                        {data.offer.promotion.title}
                    </Box>
                </ButtonBase>
            </Grid>
            <Grid item xs={11} sm container>
                <Grid item xs container direction="column" spacing={SPACING_TWO}>
                    <Grid item xs data-testid="property_title">
                        <Typography gutterBottom variant="body1" component="div">
                            <Link href="#" underline="none"
                                  color="inherit"><b>{data.property.title}</b>
                            </Link> {AccRates}
                        </Typography>
                        <Typography variant="subtitle2" gutterBottom data-testid="property_address">
                            {data.property.address.join(", ")}
                        </Typography>
                        <Typography variant="body2" sx={{color: "red"}} gutterBottom data-testid="offer_name">
                            <u>{data.offer.name}</u>
                        </Typography>
                    </Grid>
                    <Grid item>
                        <Typography variant="body2" sx={{color: "green"}} gutterBottom data-testid="cancellation">
                            {data.offer.cancellationOption.cancellationType === "FREE_CANCELLATION" && "Free cancellation"}
                        </Typography>
                    </Grid>
                </Grid>
                <Grid item>
                    <Stack alignItems="flex-end">
                        <Typography variant="subtitle1">
                            1 night total (AUD)
                        </Typography>
                        <Typography variant="h5" data-testid="price">
                            ${data.offer.displayPrice.amount}
                        </Typography>
                        {data.offer.savings && (
                            <Typography
                                variant="subtitle1"
                                sx={{color: "red"}}
                                data-testid="offer_saving"
                            >
                                Save ${data.offer.savings.amount}~
                            </Typography>
                        )}
                    </Stack>
                </Grid>
            </Grid>
        </Grid>
    );
};

export default Card;