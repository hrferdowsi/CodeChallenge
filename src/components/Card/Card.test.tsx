import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import Card from './Card';
import {cardMockData} from './cardMockData';

import * as mockData from '../../mockData/data.json';

import resetAllMocks = jest.resetAllMocks;

describe("Card Component", () => {
    beforeEach(() => {
        resetAllMocks()
    })
    const tree = mount(
        <Card data={cardMockData}/>
    )
    test('should render without crash', () => {
        expect(tree).toMatchSnapshot();
    })

    test("should contain CircleSharpIcon", () => {
        const ratingIcon = tree.find("[data-testid='rating']").find("svg")
        // expect(ratingIcon).toBeDefined()
        console.log(ratingIcon.debug())
    })

    // // manually trigger the callback
    // tree.props.onMouseEnter();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
    //
    // // manually trigger the callback
    // tree.props.onMouseLeave();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
})
