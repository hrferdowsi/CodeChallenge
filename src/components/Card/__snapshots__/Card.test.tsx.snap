// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should match the snapshot without crash 1`] = `ReactWrapper {}`;
